src/extensions/csfs.rs

```diff
@@ -0,0 +1,266 @@
+    fn corr_prop(&self) -> Correctness {
+        Correctness {
+            base: Base::B,
+            input: Input::Any, // two inputs: message and signature
+            dissatisfiable: true,
+            unit: true,
+        }
+    }
+
+    fn mall_prop(&self) -> Malleability {
+        Malleability {
+            dissat: Dissat::Unique, // multi-dissat
+            safe: true,
+            non_malleable: true,
+        }
+    }
+
+    fn extra_prop(&self) -> ExtData {
+        ExtData {
+            pk_cost: 32 + 1 + 1,   // 1 opcodes, 1 key push
+            has_free_verify: true, // free verify form. Checksigfromstack verify
+            stack_elem_count_sat: Some(2),
+            stack_elem_count_dissat: Some(2), // supply empty signature for dissatisfaction
+            max_sat_size: Some((64 + 32, 64 + 32)), // max message size can be 80, but API not supported in rust-secp-zkp no sighash type in signature
+            max_dissat_size: Some((1 + 32, 1 + 32)), // empty sig + max size message
+            timelock_info: TimelockInfo::default(),
+            exec_stack_elem_count_sat: Some(2),
+            exec_stack_elem_count_dissat: Some(2),
+            ops: OpLimits {
+                // Opcodes are really not relevant in tapscript as BIP342 removes all rules on them
+                count: 1,
+                sat: Some(0),
+                nsat: Some(0),
+            },
+        }
+    }
+
+    fn satisfy<S>(&self, sat: &S) -> Satisfaction
+    where
+        Pk: ToPublicKey,
+        S: Satisfier<Pk>,
+    {
+        let wit = match sat.lookup_csfs_sig(&self.pk) {
+            Some((msg, sig)) => Witness::Stack(vec![sig.as_ref().to_vec(), msg]),
+            None => Witness::Impossible,
+        };
+        Satisfaction {
+            stack: wit,
+            has_sig: false,
+        }
+    }
+
+    fn dissatisfy<S>(&self, _sat: &S) -> Satisfaction
+    where
+        Pk: ToPublicKey,
+        S: Satisfier<Pk>,
+    {
+        Satisfaction {
+            stack: Witness::Stack(vec![vec![], vec![]]), // empty message and empty sig
+            has_sig: false,
+        }
+    }
+
+    fn real_for_each_key<'a, F>(&'a self, pred: &mut F) -> bool
+    where
+        Pk: 'a,
+        Pk::Hash: 'a,
+        F: FnMut(ForEach<'a, Pk>) -> bool,
+    {
+        pred(ForEach::Key(&self.pk))
+    }
+
+    fn push_to_builder(&self, builder: elements::script::Builder) -> elements::script::Builder
+    where
+        Pk: ToPublicKey,
+    {
+        builder
+            .push_ms_key::<Pk, Tap>(&self.pk)
+            .push_opcode(opcodes::all::OP_CHECKSIGFROMSTACK)
+    }
+
+    fn script_size(&self) -> usize {
+        1 + 32 + 1 // opcode + key+ push
+    }
+
+    fn segwit_ctx_checks(&self) -> Result<(), miniscript::context::ScriptContextError> {
+        // New opcodes only supported in taproot context
+        Err(ScriptContextError::ExtensionError(
+            "CSFS only available in Taproot".to_string(),
+        ))
```

Comment:
In 761799269e0b5bf29a7a9fa6162a468653a5f351:

In the next commit you add this to `AllCov` and (as I understand it) the legacy covenant structure ... which is segwit only. But here you're saying it's taproot only? I'm a bit confused.

If this is a segwit thing we should remember that there is no such thing as a non-verify CSFS.

Reply:
We can add all fragments to `CovenantExt`. But the type system would not allow creation of these. 
- Miniscript<Pk, SegwitV0, CovenantExt<Pk>> when parsing csfs fragment. Similar to how to `multi_a` is disallowed. There is a test also checking this.
- `Miniscript<Pk, Tap, VerEq>` as that only available in Segwtiv0. 


Perhaps, we should rename VerEq to LegacyVerEq to clarify that this is based on old CSFS design paradigm.

Reply:
Yes, I'd like everything based on the old CSFS paradigm to have `Legacy` in the name.

And thanks for the explanation of how the type system keeps everything in bounds here. Very cool.